sinker/bulk_sinker.go

```diff
@@ -0,0 +1,309 @@
+	mapOutput := output.GetMapOutput()
+	if !mapOutput.MessageIs(dbChanges) && mapOutput.TypeUrl != "type.googleapis.com/sf.substreams.database.v1.DatabaseChanges" {
+		return fmt.Errorf("mismatched message type: trying to unmarshal unknown type %q", mapOutput.MessageName())
+	}
+
+	// We do not use UnmarshalTo here because we need to parse an older proto type and
+	// UnmarshalTo enforces the type check. So we check manually the `TypeUrl` above and we use
+	// `Unmarshal` instead which only deals with the bytes value.
+	if err := proto.Unmarshal(mapOutput.Value, dbChanges); err != nil {
+		return fmt.Errorf("unmarshal database changes: %w", err)
+	}
+
+	if err := s.dumpDatabaseChangesIntoCSV(dbChanges); err != nil {
+		return fmt.Errorf("apply database changes: %w", err)
+	}
+
+	// we set the cursor before rolling because roll process shutdown the system
+	// TODO update to use queue
+	s.stateStore.SetCursor(cursor)
+	if cursor.Block().Num()%s.bundleSize == 0 {
+		state, err := s.stateStore.GetState()
+		if err != nil {
+			s.Shutdown(fmt.Errorf("unable to get state: %w", err))
+			return fmt.Errorf("unable to get state: %w", err)
+		}
+		if err := state.Save(); err != nil {
+			s.Shutdown(fmt.Errorf("unable to save state: %w", err))
+			return fmt.Errorf("unable to save state: %w", err)
+		}
+		s.stateStore.UploadCursor(state)
+	}
+
+	s.rollAllBundlers(ctx, data.Clock.Number, cursor)
+
+	return nil
+}
+
+func (s *BulkSinker) dumpDatabaseChangesIntoCSV(dbChanges *pbdatabase.DatabaseChanges) error {
+	for _, change := range dbChanges.TableChanges {
+		if !s.loader.HasTable(change.Table) {
+			return fmt.Errorf(
+				"your Substreams sent us a change for a table named %s we don't know about on %s (available tables: %s)",
+				change.Table,
+				s.loader.GetIdentifier(),
+				s.loader.GetAvailableTablesInSchema(),
+			)
+		}
+
+		var fields map[string]string
+		switch u := change.PrimaryKey.(type) {
+		case *pbdatabase.TableChange_Pk:
+			var err error
+			fields, err = s.loader.GetPrimaryKey(change.Table, u.Pk)
+			if err != nil {
+				return err
+			}
+		case *pbdatabase.TableChange_CompositePk:
+			fields = u.CompositePk.Keys
+		default:
+			return fmt.Errorf("unknown primary key type: %T", change.PrimaryKey)
+		}
+		table := change.Table
+		tableBundler, ok := s.fileBundlers[table]
+		if !ok {
+			return fmt.Errorf("cannot get bundler writer for table %s", table)
+		}
+		switch change.Operation {
+		case pbdatabase.TableChange_CREATE:
+			// add fields
+			for _, field := range change.Fields {
+				fields[field.Name] = field.NewValue
+			}
+			data, _ := bundler.CSVEncode(fields)
+			if !tableBundler.HeaderWritten {
+				tableBundler.Writer().Write(tableBundler.Header)
+				tableBundler.HeaderWritten = true
+			}
+			tableBundler.Writer().Write(data)
+		case pbdatabase.TableChange_UPDATE:
+		case pbdatabase.TableChange_DELETE:
+		default:
+			return fmt.Errorf("Currently, we only support append only databases")
```

Comment:
Go does not do `fallthrough` by default, you need to explicit request `fallthrough` by using the keyword:
```suggestion
		case pbdatabase.TableChange_UPDATE:
		    fallthrough
		case pbdatabase.TableChange_DELETE:
		    fallthrough
		default:
			return fmt.Errorf("currently, we only support append only databases (performing only inserts)")
```

Also error should always be lowered case (and I changed the wording a bit).

Reply:
Done